File: src/test/java/edu/metrostate/trackingsystem/infrastructure/database/ResultTest.java
package edu.metrostate.trackingsystem.infrastructure.database;

import com.google.gson.JsonObject;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ResultTest {

    @Test
    void success() {
        var result = Result.success();
        assertTrue(result.isSuccess());
    }

    @Test
    void testSuccess() {
        assertTrue(Result.success().isSuccess());
    }

    @Test
    void failure() {
        assertFalse(Result.failure("").isSuccess());
    }

    @Test
    void isSuccess() {
        var result = Result.failure("");
        assertFalse(result.isSuccess());
    }

    @Test
    void getData() {
        var result = Result.success(new JsonObject());
        assertNotNull(result.getData());
    }

    @Test
    void getErrorMessage() {
        var result = Result.failure("error");
        assertEquals("error", result.getErrorMessage());
    }

    @Test
    void cantGetDataWhenFailure() {
        var result = Result.failure("error");
        assertThrows(IllegalStateException.class, result::getData);
    }
}

File: src/test/kotlin/edu/metrostate/dealership/infrastructure/database/ResultTest.kt
package edu.metrostate.dealership.infrastructure.database

import com.google.gson.JsonObject
import edu.metrostate.dealership.infrastructure.database.Result.Companion.failure
import edu.metrostate.dealership.infrastructure.database.Result.Companion.success
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

internal class ResultTest {

    @Test
    fun success() {
        val result = success<Any>()
        assertTrue(result.isSuccess)
    }

    @Test
    fun testSuccess() {
        assertTrue(success<Any>().isSuccess)
    }

    @Test
    fun failure() {
        assertFalse(failure<Any>("").isSuccess)
    }

    @Test
    fun isSuccessFlag() {
        val result = failure<Any>("")
        assertFalse(result.isSuccess)
    }

    @Test
    fun dataPresentOnSuccess() {
        val result = success(JsonObject())
        assertNotNull(result.getData())
    }

    @Test
    fun errorMessageOnFailure() {
        val result = failure<Any>("error")
        assertEquals("error", result.errorMessage)
    }

    @Test
    fun cantGetDataWhenFailure() {
        val result = failure<Any>("error")
        assertThrows(IllegalStateException::class.java) {
            result.getData()
        }
    }
}
